File: leetcode_0496/java/leetcode_0496.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;

public class leetcode_0496 {
    class Solution {
        public int[] nextGreaterElement(int[] nums1, int[] nums2) {
            Map<Integer,Integer> record = new HashMap();
            for(int i = 0;i < nums2.length;i++)
                record.put(nums2[i],i);
            int[] res = new int[nums1.length];
            int[] ans = new int[nums2.length];
            //双向链表模拟栈
            LinkedList<Integer> stack = new LinkedList();
            for(int i = nums2.length-1;i >= 0;i--){
                //如果当前元素大于栈顶元素，则一直弹出栈顶元素
                while(!stack.isEmpty() && nums2[i] >= stack.peekLast())
                    stack.removeLast();
                //将栈顶元素(大于当前元素)或-1放入结果集中
                ans[i] = (stack.isEmpty())?-1:stack.peekLast();
                //当前元素进栈
                stack.offerLast(nums2[i]);
            }
            for(int i = 0;i < nums1.length;i++)
                res[i] = ans[record.get(nums1[i])];
            return res;
        }
    }
    public static void main(String[] args) {
        int[] nums1 = {4,1,2};
        int[] nums2 = {1,3,4,2};
        Solution solution = new leetcode_0496().new Solution();
        int[] res = solution.nextGreaterElement(nums1,nums2);
        System.out.println(Arrays.toString(res));
    }
}

File: leetcode_0496/cpp/leetcode_0496.cpp
//
// Created by wangzy on 2022/3/8.
//
#include <iostream>
#include <vector>
#include <unordered_map>
#include <stack>
#include <sstream>

using namespace std;

class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        //哈希表+单调栈
        unordered_map<int,int> record;
        int i = 0;
        for(auto num:nums2)
            record[num] = i++;
        stack<int> mono_stk;
        //可以预先处理nums2使查询nums1中的每个元素在nums2中对应位置的右边第一个更大的元素时不需要再遍历，于是问题分解为两个子问题。
        //1.如何高效的计算nums2中每个元素右边的第一个更大的值(单调栈 从右往左遍历)
        //2.如何存储第1个问题的结果(采用数组存储 不存在右第一更大存-1)
        vector<int> ans(nums2.size(),0);
        for(int j = nums2.size()-1;j >= 0;j-- ){
            //如果当前元素大于栈顶元素，则一直弹出栈顶元素
            while(!mono_stk.empty() && nums2[j] >= mono_stk.top())
                mono_stk.pop();
            //将栈顶元素(大于当前元素)或-1放入结果集中
            ans[j] = (mono_stk.empty())?-1:mono_stk.top();
            //当前元素进栈
            mono_stk.push(nums2[j]);
        }
        i = 0;
        vector<int> res(nums1.size(),0);
        for(auto num:nums1)
            res[i++] = ans[record[num]];
        return res;
    }
};

template<typename T>
void printVec(std::vector<T> l){
    std::string res = "[";
    std::string tmp = "";
    std::stringstream ss;
    for(auto ele : l){
        ss << ele;
        ss >> tmp;
        res += tmp + ",";
        ss.clear();
    }
    res = res.substr(0,res.size()-1);
    res += "]";
    std::cout<<res<<std::endl;
}

int main(int argc,char* argv[]){
    vector<int> nums1 = {4,1,2};
    vector<int> nums2 = {1,3,4,2};
    vector<int> res = Solution().nextGreaterElement(nums1,nums2);
    printVec(res);
}

